classifier/firmware.go

```diff
@@ -0,0 +1,47 @@
+package classifier
+
+import (
+	bmh "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+
+	hwcc "github.com/metal3-io/hardware-classification-controller/api/v1alpha1"
+)
+
+func checkFirmware(profile *hwcc.HardwareClassification, host *bmh.BareMetalHost) bool {
+	firmwareDetails := profile.Spec.HardwareCharacteristics.Firmware
+	if firmwareDetails == nil {
+		return true
+	}
+
```

Comment:
I don't think this logic for handling empty strings vs non-empty strings is going to do the right thing. If someone specifies a `Version` but not a `Vendor` (which is supported by the API definition and is something they might reasonably do if their hardware is uniform), shouldn't we check the Version? As this is written, we won't, because line 15 will set ok to false, and we'll return early.

I think the right logic is to have a simple check for the empty string, and if a value is not empty the rule is enabled. Something like

```
if firmwareDetails.Bios.Vendor != "" {
  ok := true
  if firmwareDetails.Bios.Vendor != host.Status.HardwareDetails.Firmware.BIOS.Vendor {
    ok = false
  }
  log.Info("Firmware",
    "host", host.Name,
    "profile", profile.Name,
     "namespace", host.Namespace,
    "vendor", firmwareDetails.Bios.Vendor,
    "actualVendor", host.Status.HardwareDetails.Firmware.BIOS.Vendor,
    "ok", ok,
  )
  if !ok {
    return false
  }
}

// repeat the same structure for the version field
```

Reply:
The logic here was to make both compulsory to provide as version will be dependent of the vendor. 
But now we have made it independent.

Reply:
The new version isn't quite right either. We want to ignore empty rules, not treat them as a bad match. If the vendor string is empty, we still want to check the version.